ibis/sql/compiler.py

```diff
@@ -1493,29 +1477,29 @@ def equals(self, other, cache=None):
             cache = {}
 
         if (self, other) in cache:
-            return cache[(self, other)]
+            return cache[self, other]
 
-        if id(self) == id(other):
-            cache[(self, other)] = True
+        if self is other:
+            cache[self, other] = True
             return True
 
         if not isinstance(other, Select):
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
         this_exprs = self._all_exprs()
         other_exprs = other._all_exprs()
 
         if self.limit != other.limit:
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
         for x, y in zip(this_exprs, other_exprs):
```

Comment:
This needs to check the length of `this_exprs` and `other_exprs` otherwise we'll hit the same bug we have for equality.